core/src/main/scala/pickling/PBuilderReader.scala

```diff
@@ -16,29 +16,175 @@ trait Hintable {
+ * 6. Every putField() call must happen within a beginEntry()/endEntry() block.
+ * 7. There is no guarantee that putElement() will be called within a beginCollectoin()/endCollection() pair.
+ *    i.e. we can write empty collections.
+ * 8. There is no guarantee that putField will be called within a beginEntry()/endEntry() pair.
+ *    i.e. if we don't put any fields, this means the entry was for a "primitive" type, at least what
+ *    The pickling library considers primitives.
+ * 9. The order of putField calls in any pickler will be the exact same ordering when unpickling, if the format
+ *    is compatible.
+ *
+ * Here is a list of all types the auto-generated Picklers considers "primitives" and must be directly supported by
+ * any PBuilder:
+ *
+ *   - Nothing
+ *   - Null
+ *   - Unit
+ *   - Byte
+ *   - Char
+ *   - String
+ *   - Short
+ *   - Int
+ *   - Long
+ *   - Float
+ *   - Double
+ *   - Ref  (for circular object graphs)
+ *   - ArrayByte
+ *   - ArrayShort
+ *   - ArrayChar
+ *   - ArrayInt
+ *   - ArrayLong
+ *   - ArrayBoolean
+ *   - ArrayFloat
+ *   - ArrayDouble
+ */
 trait PBuilder extends Hintable {
+  /** Called to denote that an object is about to be serialized.
+    * @param picklee
+    *                The object to be serialized.  This may be a primtiive, in which case
+    *                it can be immediately serialized (or you can wait unitl endEntry is called).
+    * @return
+    *                A pbuilder instance a pickler can use to serialize the picklee, if it's a complex type.
+    */
   def beginEntry(picklee: Any): PBuilder
+  /**
+   * Serialize a "field" in a complex structure/object being pickled.
+   * @param name  The name of the field to serialize.
+   * @param pickler  A callback which will be passed an appropriate pickler.
+   *                 You should ensure this function will perform a beginEntry()/endEntry() block.
+   * @return A builder for remaining items in the current complex structure being pickled.
+   */
   def putField(name: String, pickler: PBuilder => Unit): PBuilder
+
+  /**
+   * Call this to denote that the given primitive, collection or structure being pickled is completed.
+   */
   def endEntry(): Unit
+
+  /**
+   * Denotes that a collection of elements is about to be pickled.
+   *
+   * Note: This must be called after beginEntry()
+   * @param length   The length of the collection being serialized.
+   * @return  A pickler which can serialzie the collection.
+   */
   def beginCollection(length: Int): PBuilder
+
+  /**
+   * Places the next element in the serialized collection.
+   *
+   * Note: This must be called after beginCollection().
+   * @param pickler  A callback which is passed a pickler able to serialize the item in the collection.
+   * @return  A pickler which can serialize the next element of the collection.
+   */
   def putElement(pickler: PBuilder => Unit): PBuilder
```

Comment:
so each call has to be on the builder returned from the last... is endCollection called on the returned builder from the last putElement, or on the same builder used to call beginCollection, or on the builder returned from beginCollection?